From 5dc3c211163bffdc734f88626b5999493cb51d5e Mon Sep 17 00:00:00 2001
From: Nick DeLuca <nickdeluca08@gmail.com>
Date: Fri, 16 Dec 2022 21:38:37 +0000
Subject: [PATCH 11/12] patch rpc to include finalized support; if block height
 less than last upgrade then only set params that exist (skip the eip712
 allowed list to prevent panics on eth_getCall)

---
 rpc/ethereum/types/block.go | 11 ++++++-----
 x/evm/keeper/params.go      |  6 +++++-
 2 files changed, 11 insertions(+), 6 deletions(-)

diff --git a/rpc/ethereum/types/block.go b/rpc/ethereum/types/block.go
index 0f407f90..92a455eb 100644
--- a/rpc/ethereum/types/block.go
+++ b/rpc/ethereum/types/block.go
@@ -30,9 +30,10 @@ const (
 )
 
 const (
-	BlockParamEarliest = "earliest"
-	BlockParamLatest   = "latest"
-	BlockParamPending  = "pending"
+	BlockParamEarliest  = "earliest"
+	BlockParamLatest    = "latest"
+	BlockParamFinalized = "finalized"
+	BlockParamPending   = "pending"
 )
 
 // NewBlockNumber creates a new BlockNumber instance.
@@ -72,7 +73,7 @@ func (bn *BlockNumber) UnmarshalJSON(data []byte) error {
 	case BlockParamEarliest:
 		*bn = EthEarliestBlockNumber
 		return nil
-	case BlockParamLatest:
+	case BlockParamLatest, BlockParamFinalized:
 		*bn = EthLatestBlockNumber
 		return nil
 	case BlockParamPending:
@@ -158,7 +159,7 @@ func (bnh *BlockNumberOrHash) decodeFromString(input string) error {
 	case BlockParamEarliest:
 		bn := EthEarliestBlockNumber
 		bnh.BlockNumber = &bn
-	case BlockParamLatest:
+	case BlockParamLatest, BlockParamFinalized:
 		bn := EthLatestBlockNumber
 		bnh.BlockNumber = &bn
 	case BlockParamPending:
diff --git a/x/evm/keeper/params.go b/x/evm/keeper/params.go
index dbf0a0fc..3d5a21e9 100644
--- a/x/evm/keeper/params.go
+++ b/x/evm/keeper/params.go
@@ -8,7 +8,11 @@ import (
 
 // GetParams returns the total set of evm parameters.
 func (k Keeper) GetParams(ctx sdk.Context) (params types.Params) {
-	k.paramSpace.GetParamSet(ctx, &params)
+	if ctx.BlockHeight() < 2098400 {
+		k.paramSpace.GetParamSetIfExists(ctx, &params)
+	} else {
+		k.paramSpace.GetParamSet(ctx, &params)
+	}
 	return params
 }
 
-- 
2.38.1

